test/unit/forge/routes/api/project_spec.js

```diff
@@ -755,222 +755,410 @@ describe('Project API', function () {
 
-            const projectType = {
-                name: 'projectType2',
-                description: 'default project type',
-                active: true,
-                properties: { foo: 'bar' },
-                order: 2
-            }
-            const projectType2 = await app.db.models.ProjectType.create(projectType)
+                // Create a new stack
+                const stackProperties = {
+                    name: 'stack-new',
+                    active: true,
+                    properties: { nodered: '9.9.9' }
+                }
+                const stack = await app.db.models.ProjectStack.create(stackProperties)
+                await stack.setProjectType(projectType)
 
-            const response = await app.inject({
-                method: 'PUT',
-                url: `/api/v1/projects/${project2.id}`,
-                payload: {
-                    projectType: projectType2.hashid
-                },
-                cookies: { sid: TestObjects.tokens.alice }
-            })
-            response.statusCode.should.equal(400)
-        })
-        it('Can change project-type if not set', async function () {
-            const project2 = await app.db.models.Project.create({ name: 'project2', type: '', url: '' })
-            await TestObjects.ATeam.addProject(project2)
-            await project2.setProjectStack(TestObjects.stack1)
-            await project2.setProjectTemplate(TestObjects.template1)
+                // Put project in running state
+                await app.containers.start(project)
 
-            const response = await app.inject({
-                method: 'PUT',
-                url: `/api/v1/projects/${project2.id}`,
-                payload: {
-                    projectType: TestObjects.projectType1.hashid
-                },
-                cookies: { sid: TestObjects.tokens.alice }
-            })
-            response.statusCode.should.equal(200)
-        })
+                const response = await app.inject({
+                    method: 'PUT',
+                    url: `/api/v1/projects/${TestObjects.project1.id}`,
+                    payload: {
+                        projectType: projectType.id,
+                        stack: stack.id,
+                        changeProjectDefinition: true
+                    },
+                    cookies: { sid: TestObjects.tokens.alice }
+                })
 
-        it('Change project stack', async function () {
-            // Setup some flows/credentials
-            await addFlowsToProject(TestObjects.project1.id,
-                TestObjects.tokens.project,
-                [{ id: 'node1' }],
-                { testCreds: 'abc' },
-                'key1',
-                {
-                    httpAdminRoot: '/test-red',
-                    env: [
-                        { name: 'one', value: 'a' },
-                        { name: 'two', value: 'b' }
+                response.statusCode.should.equal(200)
```

Comment:
Unfortunately, I couldn't find a way to assert the project is suspended as the request only returns once the project has been restarted.

Reply:
One approach would be to add some sort of internal logging to the stub driver that can be used to check what has been done to a given project. But not sure that's worth the effort at this point.

Reply:
We could spy on the audit logs actually.

Reply:
I did wonder that, but couldn't recall if the audit logging is done at that level.